pkg/convert/pprof/streaming/labels_cache.go

```diff
@@ -0,0 +1,116 @@
+package streaming
+
+import (
+	"encoding/binary"
+	"github.com/cespare/xxhash/v2"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/tree"
+	"sort"
+)
+
+type Labels []label
+
+func (l Labels) Len() int           { return len(l) }
+func (l Labels) Less(i, j int) bool { return l[i].k < l[j].k }
+func (l Labels) Swap(i, j int)      { l[i], l[j] = l[j], l[i] }
+
+func (l Labels) Hash() uint64 {
+	h := xxhash.New()
+	var t [16]byte
+	sort.Sort(l)
+	for _, x := range l {
+		if x.v == 0 {
+			continue
+		}
+		binary.LittleEndian.PutUint64(t[0:8], uint64(x.k))
+		binary.LittleEndian.PutUint64(t[8:16], uint64(x.v))
+		_, _ = h.Write(t[:])
+	}
+	return h.Sum64()
+}
+
+// sample type -> labels hash -> entry
+
+type LabelsCache map[int64]map[uint64]*LabelsCacheEntry
```

Comment:
In future we may want to optimise the structure – this is the hottest path.

We know that there aren't this many sample types and the set is dense, therefore we don't need the top level map, it should be replaces with a slice
```go
[]map[uint64]*LabelsCacheEntry
```

Next, `*LabelsCacheEntry` is actually a tree and associated Labels. We can represent them in Structure of Arrays way:

```go
type cache struct {
   // sample type -> labels hash -> index to labels and trees
   indices []map[uint64]int
   labels  []Labels
   trees   []*Tree
}
```

Another possible improvement is getting rid of Labels in favour of an arena:

```go
type cache struct {
   // sample type -> labels hash -> index to labelRefs and trees
   indices []map[uint64]int
   // A label reference points to the subset in labels:
   // Hight 32 bits is the start offset, lower 32 bits is the subset size.
   labelRefs []uint64
   labels    []Label
   trees     []*Tree
}
```

This allows reusing the cache easily. Also, `map[int]int can be optimised further (e.g https://github.com/brentp/intintmap).

Reply:
after our call I had an impression that with a new cloudstorage's ingesters we will not need the LabelsCache at all so I did not even try to optimize it

Reply:
Honestly, I'm not 100% sure when (if) it will be released as part of the OSS version. Also, the cache is only needed to do diffs of cumulative profiles, so I think we can't remove it completely :/ Moving the logic to the storage is kinda problematic, though possible I guess